repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneRequestHandler.scala

```diff
@@ -0,0 +1,313 @@
+        )(Endpoint.apply _)
+
+      val json = Json.parse(jsonString)
+      //Have to convert it to a vector, because List isn't serializeable in 2.10
+      (json \ "endpoints").validate[Vector[Endpoint]] match {
+        case s: JsSuccess[Vector[Endpoint]] =>
+          val endpoints = s.get
+          Success(new EndpointsData(jsonString, endpoints))
+        case f: JsError =>
+          Failure(IdentityCommunicationException("Identity didn't respond with proper Endpoints JSON"))
+      }
+    }
+
+    val akkaResponse = Try(akkaServiceClient.get(s"$ENDPOINTS_KEY_PREFIX$forToken",
+      s"$identityServiceUri${ENDPOINTS_ENDPOINT(forToken)}",
+      (Map(CommonHttpHeader.AUTH_TOKEN.toString -> authenticatingToken,
+        CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+        ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString -> _)).asJava))
+
+    akkaResponse match {
+      case Success(serviceClientResponse) =>
+        serviceClientResponse.getStatus match {
+          case SC_OK | SC_NON_AUTHORITATIVE_INFORMATION =>
+            val jsonResponse = Source.fromInputStream(serviceClientResponse.getData).getLines mkString ""
+            extractEndpointInfo(jsonResponse)
+          case SC_UNAUTHORIZED =>
+            Failure(AdminTokenUnauthorizedException("Admin token unauthorized to access endpoints"))
+          case SC_FORBIDDEN => Failure(IdentityAdminTokenException("Admin token forbidden from accessing endpoints"))
+          case SC_NOT_FOUND => Failure(InvalidTokenException("Provided token went bad before endpoints call"))
+          case SC_REQUEST_ENTITY_TOO_LARGE | SC_TOO_MANY_REQUESTS =>
+            Failure(OverLimitException(buildRetryValue(serviceClientResponse), "Rate limited when accessing endpoints"))
+          case statusCode if statusCode >= 500 =>
+            Failure(IdentityCommunicationException("Identity Service not available to get endpoints"))
+          case _ => Failure(new Exception("Unexpected response code from the endpoints call"))
+        }
+      case Failure(x) => Failure(new Exception("Failure communicating with identity during get endpoints", x))
+    }
+  }
+
+  final def getGroups(authenticatingToken: String, forToken: String): Try[Vector[String]] = {
+    def extractGroupInfo(inputStream: InputStream): Try[Vector[String]] = {
+      Try {
+        val input: String = Source.fromInputStream(inputStream).getLines mkString ""
+        val json = Json.parse(input)
+
+        (json \ "RAX-KSGRP:groups" \\ "id").map(_.as[String]).toVector
+      }
+    }
+
+    val akkaResponse = Try(akkaServiceClient.get(s"$GROUPS_KEY_PREFIX$forToken",
+      s"$identityServiceUri${GROUPS_ENDPOINT(forToken)}",
+      (Map(CommonHttpHeader.AUTH_TOKEN.toString -> authenticatingToken,
+        CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+        ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString -> _)).asJava))
+
+    akkaResponse match {
```

Comment:
you have this block three times where the only difference is error message strings and what to do on success. I'd super love it if it could be condensed.